msgpack-core/src/main/java/org/msgpack/core/buffer/MessageBuffer.java

```diff
@@ -59,101 +44,62 @@
-                    constructorType = DirectBufferConstructorType.ARGS_LONG_INT;
-                } catch (NoSuchMethodException e1) {
-                    try {
-                        directByteBufferConstructor = directByteBufferClass.getDeclaredConstructor(int.class, int.class);
-                        constructorType = DirectBufferConstructorType.ARGS_INT_INT;
-                    } catch (NoSuchMethodException e2) {
-                        Class<?> aClass = Class.forName("java.nio.MemoryBlock");
-                        mbWrap = aClass.getDeclaredMethod("wrapFromJni", int.class, long.class);
-                        mbWrap.setAccessible(true);
-                        directByteBufferConstructor = directByteBufferClass.getDeclaredConstructor(aClass, int.class, int.class);
-                        constructorType = DirectBufferConstructorType.ARGS_MB_INT_INT;
-                    }
-                }
-            }
-            byteBufferConstructor = directByteBufferConstructor;
-            directBufferConstructorType = constructorType;
-            memoryBlockWrapFromJni = mbWrap;
-            if(byteBufferConstructor == null)
-                throw new RuntimeException("Constructor of DirectByteBuffer is not found");
-
-            byteBufferConstructor.setAccessible(true);
-
-            // Check the endian of this CPU
-            boolean isLittleEndian = true;
-            byte[] a = new byte[8];
-            // use Unsafe.putLong to an array since Unsafe.getByte is not available in Android
-            unsafe.putLong(a, (long) ARRAY_BYTE_BASE_OFFSET, 0x0102030405060708L);
-            switch(a[0]) {
-                case 0x01:
-                    isLittleEndian = false;
-                    break;
-                case 0x08:
-                    isLittleEndian = true;
-                    break;
-                default:
-                    assert false;
-            }
+            // For Java6, android and JVM that has no Unsafe class, use Universal MessageBuffer
+            isUniversalBuffer =
+                Boolean.parseBoolean(System.getProperty("msgpack.universal-buffer", "false"))
+                    || isAndroid
+                    || isGAE
+                    || !isJavaAtLeast7
+                    || !hasUnsafe;
 
             // We need to use reflection to find MessageBuffer implementation classes because
             // importing these classes creates TypeProfile and adds some overhead to method calls.
             String bufferClsName;
-            boolean useUniversalBuffer = Boolean.parseBoolean(System.getProperty("msgpack.universal-buffer", "false"));
-            if(!useUniversalBuffer && !isAndroid && isJavaAtLeast7) {
-                if(isLittleEndian)
-                    bufferClsName = "org.msgpack.core.buffer.MessageBuffer";
-                else
-                    bufferClsName = "org.msgpack.core.buffer.MessageBufferBE";
-            }
-            else {
+            if(isUniversalBuffer) {
```

Comment:
Good idea. With this fix, msgpack-java works anyway. To do this, I need to put the initialization code of static fields (bufferClsName, unsafe, etc.) after the last catch call.